### PR TITLE
Fortran implementation of gyro correction

### PR DESCRIPTION
**Pull request by JETanner**
This is the first draft to implement the gyro correction in process

### CODE REVIEW

File: src/fidasim.f90 [hunk 1]
```diff
@@ -2301,6 +2385,19 @@ subroutine read_equilibrium
...
+        call deriv(inter_grid%r, equil%fields(:,iz)%bt, equil%fields(:,iz)%dbt_dr)
+        call deriv(inter_grid%r, equil%fields(:,iz)%bz, equil%fields(:,iz)%dbz_dr)
+    enddo
+
+    do ir=1,inter_grid%nr
+        call deriv(inter_grid%z, equil%fields(ir,:)%br, equil%fields(ir,:)%dbr_dz)
+        call deriv(inter_grid%z, equil%fields(ir,:)%bt, equil%fields(ir,:)%dbt_dz)
+        call deriv(inter_grid%z, equil%fields(ir,:)%bz, equil%fields(ir,:)%dbz_dz)
+    enddo
+
```
Comment (lstagner) [inline src/fidasim.f90]: You have a routine to calculate derivatives in two dimensions why not use those?

File: src/fidasim.f90 [hunk 2]
```diff
@@ -6117,8 +6217,28 @@ subroutine gyro_step(vi, fields, r_gyro)
...
+        grad_B(2) = 0.0
+        grad_B(3) = (fields%br * fields%dbz_dr + fields%dbt_dz + fields%bz*fields%dbz_dz) /&
+                    fields%b_abs
+        rg_rtz(1) = rg_uvw(1) * cos(phi) + rg_uvw(2) * sin(phi)
+        rg_rtz(2) = 0.0
+        rg_rtz(3) = rg_uvw(3)
+        term2 = -1.0 / (2.0 * fields%b_abs) * dot_product(rg_rtz,grad_B)
+        r_gyro = r_gyro * (1.0 - term1 - term2)
+        if(1.0 - term1 - term2 .le. 0.0) then
+            write(*,'(a)') '---- Error in subroutine gyro_step ----'
```
Comment (lstagner) [inline src/fidasim.f90]: Could you change the error message to have the same format as the other messages in the code.

File: src/fidasim.f90 [hunk 3]
```diff
@@ -6117,8 +6217,28 @@ subroutine gyro_step(vi, fields, r_gyro)
         else
             rg_uvw = matmul(beam_grid%basis,r_gyro)
         endif
-        rg_r = rg_uvw(1)*cos(phi) + rg_uvw(2)*sin(phi)
-        r_gyro = r_gyro*(1 - rg_r/(2*R))
+        rg_r = rg_uvw(1) * cos(phi) + rg_uvw(2) * sin(phi)
```
Comment (lstagner) [inline src/fidasim.f90]: General style rule: Don't put spaces between `*` and `/` only for `+` and `-`. It looks cleaner that way

File: src/fidasim.f90 [hunk 4]
```diff
@@ -6094,13 +6191,16 @@ subroutine gyro_step(vi, fields, r_gyro)
         !+ Electro-magnetic fields
     real(Float64), dimension(3), intent(out) :: r_gyro
         !+ Gyro-step
+        !+ Gyro-radius vector from partile position to guiding center
+        !+ vec\tho = (\hat b cross \vec v)/Omega
```
Comment (lstagner) [inline src/fidasim.f90]: misspelled `particle`. Also just get rid of the latex.

File: src/utilities.f90 [hunk 5]
```diff
@@ -500,4 +507,71 @@ function get_value(SA, subs) result (val)
 
 end function get_value
 
+!============================================================================
+!--------------------------------Deriv Routines------------------------------
+!============================================================================
+
+subroutine deriv_1d(x,y,yp)
```
Comment (lstagner) [inline src/utilities.f90]: You have an error in this routine.

**Comment by lstagner** [comment]
image

Does that look right to you?

**Comment by JETanner** [comment]
Yeah I got something on the last point running this again just now but it
did not show in the first test I am looking into it. Also I don't know if
an error I am seeing is because something is not installed right (I am
getting a lot of errors for things I did not have before - like is there
anything I am missing to make Float64 under the real declaration not work?)

Joshua

**Comment by JETanner** [comment]
fixed it, reuploaded, didn't notice it in the test since I apparently did
not graph the last point (I don't know how to graph directly from fortran
so I had saved it to a text file and just grabbed it  and put it into excel
via copy and paste a large list of values. (I thought all of them but
apparently clipboard only let me grab like 70% of the text)

Joshua

File: src/utilities.f90 [hunk 6]
```diff
@@ -500,4 +507,71 @@ function get_value(SA, subs) result (val)
...
+        !+ Z Values
+    real(Float64), dimension(:,:), intent(out) :: zxp
+        !+ Derivative of Z w.r.t. X
+    real(Float64), dimension(:,:), intent(out) :: zyp
+        !+ Derivative of Z w.r.t. Y
+
+    integer :: i,n
+        !! temporary values for loops
+
+    n = size(x)
```
Comment (lstagner) [inline src/utilities.f90]: I think this should be `n = size(y)`

File: src/utilities.f90 [hunk 7]
```diff
@@ -500,4 +507,71 @@ function get_value(SA, subs) result (val)
...
+    do i = 1,n
+        call deriv_1d(x,z(:,i),zxp(:,i))
+    enddo
+
+    n = size(y)
```
Comment (lstagner) [inline src/utilities.f90]: `n=size(x)`

**Comment by JETanner** [comment]
Made corrections and latest test run got expected corrections:
jt_gyro

**Comment by lstagner** [comment]
Can you also include a plot comparing yours version with Deyongs

**Comment by JETanner** [comment]
Here is the graph with Deyongs values added for comparision
both

**Comment by lstagner** [comment]
Merged locally.

**Pull request closed without merging**